### PR TITLE
feat: Add support for declare, deploy, and invoke v3. Also add support for estimate_fee

### PR DESCRIPTION
**Pull request by rukafe0x**
All tested locally in devnet 0.1.2 with "melos test" and "act -j lint".

(close issues #339 #340 #341 #342)

## Summary by CodeRabbit

- **New Features**
	- Introduced a new smart contract for a counter with methods for incrementing and retrieving the current count.
	- Added an ERC20 token contract implementation with functionality for transfers and approvals.
	- Enhanced transaction handling with support for new versions and additional parameters for fee estimation.
	- Added new constants and updated existing ones for better transaction version management.
	- Introduced a new transaction type, `BroadcastedInvokeTxnV3`, with additional fields for fee estimation.
	- Expanded the `Account` class to support new transaction version (v3) with enhanced fee estimation capabilities.
	- Updated the `execute` method in the `Contract` class for improved transaction configurations.
	- Enhanced the `FeeEstimate` class to include additional metrics for gas consumption and pricing.
	- Added new JSON files for compiled contract data and contract class definitions.
	- Updated transaction models to support new versions and fields, including `DeclareTransactionV3` and `DeployAccountTransactionV3`.

- **Bug Fixes**
	- Improved error handling in transaction tests and fee estimation processes.

- **Documentation**
	- Updated comments and documentation across various files for clarity and consistency.

- **Tests**
	- Expanded test coverage for new transaction types and functionalities, including fee estimation for different versions.

### CODE REVIEW

File: packages/starknet_provider/lib/src/model/fee_estimate.freezed.dart [hunk 1]
```diff
@@ -21,8 +21,11 @@ FeeEstimate _$FeeEstimateFromJson(Map<String, dynamic> json) {
 /// @nodoc
 mixin _$FeeEstimate {
   String get gasConsumed => throw _privateConstructorUsedError;
+  String get dataGasConsumed => throw _privateConstructorUsedError;
   String get gasPrice => throw _privateConstructorUsedError;
+  String get dataGasPrice => throw _privateConstructorUsedError;
   String get overallFee => throw _privateConstructorUsedError;
+  String get unit => throw _privateConstructorUsedError;
```
[the remark this answers is not in the document]
Comment (rukafe0x) [reply packages/starknet_provider/lib/src/model/fee_estimate.freezed.dart:28]: This is a self-generated file, we can't change it.

File: contracts/v2.6.2/src/erc20.cairo [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    fn constructor(
+        ref self: ContractState,
+        initial_supply: u256,
+        recipient: ContractAddress
+    ) {
+        let name = "MyToken";
+        let symbol = "MTK";
+
+        self.erc20.initializer(name, symbol);
+        self.erc20._mint(recipient, initial_supply);
```
[the remark this answers is not in the document]
Comment (rukafe0x) [reply contracts/v2.6.2/src/erc20.cairo:39]: These are the correct names for the OZ implementation.

File: packages/starknet/lib/src/types/felt.dart [hunk 3]
```diff
@@ -8,8 +8,62 @@
...
+      throw ArgumentError('Value must be greater than 0');
+    }
+    return result;
+  }
+
+  //define * operator
+  Felt operator *(Felt other) {
+    Felt result = Felt(_bigInt * other._bigInt);
+    return result;
+  }
```
[the remark this answers is not in the document]
Comment (rukafe0x) [reply packages/starknet/lib/src/types/felt.dart:54]: is not required, values are validated in the felt constructor.

File: packages/starknet/lib/src/types/felt.dart [hunk 4]
```diff
@@ -8,8 +8,62 @@
...
+  //define / operator
+  Felt operator /(Felt other) {
+    if (other._bigInt == BigInt.zero) {
+      throw ArgumentError('Cannot divide by zero');
+    }
+    // modInverse will throw if no inverse exists
+    BigInt inverse = other._bigInt.modInverse(prime);
+    Felt result = Felt((_bigInt * inverse) % prime);
+    return result;
+  }
```
[the remark this answers is not in the document]
Comment (rukafe0x) [reply packages/starknet/lib/src/types/felt.dart:65]: This is not required at this time.

File: packages/starknet/lib/src/types/felt.dart [hunk 5]
```diff
@@ -8,8 +8,62 @@
...
   late BigInt _bigInt;
 
+  //define << and >> . Check result not greater or equel to prime
+  Felt operator <<(int shift) {
+    Felt result = Felt(_bigInt << shift);
+    if (result._bigInt >= prime) {
+      throw ArgumentError('Value must be smaller than 2^251 + 17 * 2^192 + 1');
+    }
+    return result;
+  }
```
[the remark this answers is not in the document]
Comment (rukafe0x) [reply packages/starknet/lib/src/types/felt.dart:22]: This is not necessary, the values are validated in the Felt constructor.

File: packages/starknet_provider/lib/src/model/deploy_account_transaction.dart [hunk 6]
```diff
@@ -7,11 +7,22 @@ import 'package:starknet_provider/starknet_provider.dart';
...
-          : throw Exception("Unsupported version ${json['version']}");
+      switch (json['version']) {
+        deployAccountTxnV1 ||
+        deployAccountTxnV1OldCompat =>
+          DeployAccountTransactionV1.fromJson(json),
+        deployAccountTxnV3 ||
+        deployAccountTxnV3OldCompat =>
+          DeployAccountTransactionV3.fromJson(json),
+        _ => throw Exception("Unsupported version ${json['version']}"),
+      };
```
[the remark this answers is not in the document]
Comment (rukafe0x) [reply packages/starknet_provider/lib/src/model/deploy_account_transaction.dart]: ok, fixed

File: packages/starknet/lib/src/account.dart [hunk 7]
```diff
@@ -508,6 +939,38 @@
...
+    // https://pub.dev/packages/xrandom
+    final rand = Random.secure();
+    final bytes = [for (int i = 0; i < 32; i++) rand.nextInt(256)];
+    final randomBigInt = bytes.fold<BigInt>(
+      BigInt.zero,
+      (prev, byte) => (prev << 8) | BigInt.from(byte),
+    );
+    final salt = Felt(randomBigInt % Felt.prime);
+    return salt;
+  }
```
[the remark this answers is not in the document]
Comment (rukafe0x) [reply packages/starknet/lib/src/account.dart:973]: this is a TODO for future versions as mentioned in comments

File: packages/starknet/lib/src/account.dart [hunk 8]
```diff
@@ -13,6 +14,22 @@ enum AccountSupportedTxVersion {
   @Deprecated('Transaction version 0 will be removed with Starknet alpha v0.11')
   v0,
   v1,
+  v3,
+}
+
```
[the remark this answers is not in the document]
Comment (rukafe0x) [reply packages/starknet/lib/src/account.dart:19]: This is not required at this time.

File: packages/starknet/lib/src/types/felt.dart [hunk 9]
```diff
@@ -8,8 +8,62 @@ class Felt {
...
+
+  Felt operator >>(int shift) {
+    Felt result = Felt(_bigInt >> shift);
+    if (result._bigInt >= prime) {
+      throw ArgumentError('Value must be smaller than 2^251 + 17 * 2^192 + 1');
+    }
+    return result;
+  }
```
[the remark this answers is not in the document]
Comment (rukafe0x) [reply packages/starknet/lib/src/types/felt.dart:30]: valid values are validated in the Felt constructor.

File: packages/starknet_provider/lib/src/model/deploy_account_transaction.dart [hunk 10]
```diff
@@ -27,14 +38,38 @@ class DeployAccountTransactionV1
...
+  const factory DeployAccountTransactionV3({
+    @Default('DEPLOY_ACCOUNT') String type,
+    @Default(deployAccountTxnV3) String version,
+    required Felt classHash,
+    required List<Felt> constructorCalldata,
+    required Felt contractAddressSalt,
+    required String feeDataAvailabilityMode,
+    required Felt nonce,
+    required String nonceDataAvailabilityMode,
+    required List<Felt> paymasterData,
```
[the remark this answers is not in the document]
Comment (rukafe0x) [reply packages/starknet_provider/lib/src/model/deploy_account_transaction.dart:63]: This is not required at this time.

File: packages/starknet_provider/lib/src/model/deploy_account_transaction.dart [hunk 11]
```diff
@@ -27,14 +38,38 @@
...
+    required Map<String, ResourceBounds> resourceBounds,
```
[the remark this answers is not in the document]
Comment (rukafe0x) [reply packages/starknet_provider/lib/src/model/deploy_account_transaction.dart:64]: It's imported through Starknet Package.

File: packages/starknet/test/account_test.dart [hunk 12]
```diff
@@ -98,138 +396,230 @@ void main() {
...
+            transactionHash: txHash!,
+            provider: account3.provider,
+          );
+        } catch (e) {
+          print(e.toString());
+          if (!e.toString().contains('Contract error')) {
+            // If already declared just continue
+            rethrow;
+          }
+        }
```
[the remark this answers is not in the document]
Comment (rukafe0x) [reply packages/starknet/test/account_test.dart]: That's okay, that's the only message we need to catch right now.

**Review by ptisserand — APPROVED** [review]
Thanks a lot for this great work!

**Pull request merged**